docs: Nginx with SSL and HTTP/2 is now default

RanadeepPolavarapu opened the pull request:
Resolves #3797.

<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
Updated the docs to remove the older Nginx non-SSL setup. Made Nginx with SSL and HTTP/2 the default. Updated all links to point to this.

Added TLS v1.3 and HTTP/2 Server Push support into the Nginx config.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
Fixes #3797.

mattermod commented:
Hello @RanadeepPolavarapu,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

Per the [Mattermost Contribution Guide](https://developers.mattermost.com/contribute/getting-started/), we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**
Once you have signed the CLA, please comment with `/check-cla` and confirm that the CLA check is green.

This is a standard procedure for many open source projects.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

RanadeepPolavarapu commented:
/check-cla

RanadeepPolavarapu commented:
/check-cla

justinegeffen commented:
Hi @RanadeepPolavarapu! Thank you for this, it's incredibly helpful! 

Before I review the docs, I'd like to propose that the new content be added to the existing page as opposed to that page being deleted and a new one created. The reason is that if the current page's link is used anywhere else, readers visiting the page once it's deleted will land on a dead-end. Usually we'd add a blurb such as "this page has been moved to..." however, in this case, the page is nested within the larger guide so it may be a jarring experience. 

I am very open to a deeper discussion and on my side will also look into a redirect - although I don't know for certain that it'll work within the structure of the content. If you have any suggestions they are very welcome! :)

RanadeepPolavarapu commented:
Hi @justinegeffen,

That's a great point. I have added the original nginx instructions back and added the HTTP/2 SSL config I made on to that document. Now, both documents point to a fully HTTP/2 enabled SSL config.

I have taken the liberty to add TLS v1.3 and HTTP/2 server push support example in the config as well in the documentation to make the config ready for the HTTP/3 QUIC changes about to launch soon in most modern browsers.

Please feel free to let me know if you wish to have something changed.

Best Regards.

justinegeffen commented:
Thanks so much, @RanadeepPolavarapu, that was really quick! I'll have a chance to review it tomorrow and will touch base with you then. :)

justinegeffen commented on source/install/config-proxy-nginx.rst:
```diff
@@ -11,13 +11,18 @@ NGINX is configured using a file in the ``/etc/nginx/sites-available`` directory
+SSL and HTTP/2 with server push are enabled in the provided configuration example.
+
+.. note::
+  You will need get valid SSL certificates for nginx to pin the certificates properly and your browser to accept the certificate as a valid CA signed certificate. If you need an example on full configuration with pinning *Let's Encrypt*, please see `Nginx HTTP/2 & SSL full configuration guide <https://docs.mattermost.com/install/config-ssl-http2-nginx.html>`__.
```
```suggestion
  You will need valid SSL certificates in order for NGINX to pin the certificates properly. Additionally, your browser must have permissions to accept the certificate as a valid CA signed certificate. If you need an example on full configuration with pinning Let's Encrypt, please see the `Nginx HTTP/2 & SSL full configuration guide <https://docs.mattermost.com/install/config-ssl-http2-nginx.html>`__.
```

justinegeffen commented on source/install/config-ssl-http2-nginx.rst:
```diff
@@ -56,53 +58,111 @@ You can use any certificate that you want, but these instructions show you how t
-       server_name   {domain-name} ;
-       return 301 https://$server_name$request_uri;
+      listen 80 default_server;
+      server_name   mattermost.example.com;
```
```suggestion
      server_name mattermost.example.com;
```

justinegeffen commented on source/install/config-ssl-http2-nginx.rst:
```diff
@@ -112,15 +172,95 @@ You can use any certificate that you want, but these instructions show you how t
+
+**Why are Websocket connections returning a 403 error?**
+
+This is likely due to a failing cross-origin check. A check is applied for WebSocket code to see if the ``Origin`` header is the same as the host header. If it's not, a 403 error is returned. Open the file ``/etc/nginx/sites-available/mattermost`` as root in a text editor and make sure that the host header being set in the proxy is dynamic:
```
```suggestion
This is likely due to a failing cross-origin check. A check is applied for WebSocket code to see if the ``Origin`` header is the same as the host header. If it's not, a 403 error is returned. Open the file ``/etc/nginx/sites-available/mattermost`` as *root* in a text editor and make sure that the host header being set in the proxy is dynamic:
```

justinegeffen commented:
Minor suggestion but otherwise good to go on my side. Thank you!

justinegeffen commented:
Adding @DSchalla for visibility.

justinegeffen commented:
@DSchalla, does this need a security review or can it be merged? :)

DSchalla commented:
Looks good from my end, thank you @RanadeepPolavarapu! I think having Sven's review before merging would make sense to take a look from a support angle.

DSchalla commented on source/install/config-ssl-http2-nginx.rst:
```diff
@@ -56,53 +58,111 @@ You can use any certificate that you want, but these instructions show you how t
+       }
+    }
+
+    map $ssl_early_data $tls1_3_early_data {
```
@RanadeepPolavarapu Curious why you use a map instead of just use `proxy_set_header Early-Data $ssl_early_data;` directly?

RanadeepPolavarapu replied:
@DSchalla oh I used this for my production server for debugging but yes. This block can be removed for the direct variable.

justinegeffen replied:
Wondering if we should note that in the code and provide the variable too?

RanadeepPolavarapu replied:
Sure will do, I will add a comment noting this.

justinegeffen commented on source/install/config-proxy-nginx.rst:
```diff
@@ -72,6 +111,14 @@ On RHEL 7 and 8, open the file ``/etc/nginx/conf.d/mattermost``.
     }
 
+    # This block is useful for debugging TLS v1.3. Please feel free to remove this
+    # and use the `$ssl_early_data` variable exposed by nginx directly should you
```
```suggestion
    # and use the `$ssl_early_data` variable exposed by Nginx directly should you
```

svelle reviewed (approved):
LGTM
